src/rpp/concurrent_queue.h

```diff
@@ -22,7 +23,7 @@ namespace rpp
      * @note This is not optimized for speed, but has acceptable performance
      *       and due to its simplicity it won't randomly deadlock on you.
      */
-    template<class T>
+    template<class T, class mutex_t = rpp::mutex, class cond_var_t = rpp::condition_variable>
```

Comment:
should instead replace with a singular locking strategy pattern
```cpp
template<class T, class sync_t = rpp::mutex_sync>
class concurrent_queue
{
    using mutex_t = typename sync_t::mutex_t;
    uinsg cond_var_t = typename sync_t::cond_var_t;
    using lock_t = std::unique_lock<mutex_t>;

    std::unique_ptr<sync_t> DefaultSync;
    sync_t& Sync;

public:
    concurrent_queue()
        : DefaultSync{std::make_unique<sync_t>())
        , Sync{*DefaultSync}
    {}

    concurrent_queue(sync_t& sync) noexcept
        : Sync{sync}
    {}
};
```
I'm not sure which is more optimal, unique ptr or std::optional, needs to be investigated